### ISSUE TITLE
DeleteOlderThanDays

### ISSUE THREAD

**Issue by AlfredHSwe** [FSLogix/Invoke-FslShrinkDisk#31]
Hi,
I´ve just tried your script and the shrink function works perfect.
But I cant seem to delete VHDX files that are really old.
When I run 
.\Invoke-FslShrinkDisk.ps1 -Path "path to profile" -DeleteOlderThanDays 30
Nothing happens. The profile was last used in 2020-06-11.
What am I missing? 

Thanks in advance

**Comment by StevenNoel** [comment]
experiencing the same thing.  We run this script weekly and set the 'deleteolderthandays' to 90....however each week when the script runs, it changes the VHDX modified date, thus never reaching the 'deleteolderthandays' value.  Was thinking of looking at last modified date on ntuser.dat or .ost file, but since it's not mapping a drive, I'm not sure what the best way would be.  Anyone have any thoughts?

**Comment by MrFly72** [comment]
There would be an easy solution for this:
$FileWriteTimeUTCBeforeShrink=(Get-Item -Path $VHDXFileName).LastWriteTimeUTC
Then after the shrink:
$FileInfosAfterShrink=Get-Item -Path $VHDXFileName
$FileInfosAfterShrink.LastWriteTimeUTC=$FileWriteTimeUTCBeforeShrink
Did not look at the script thoroughly, but I believe that the script is reading LastWriteTime/UTC.
I am right before my yearly holiday, so no time to look more than this.

**Comment by RyanSwanepoel** [comment]
We are experiencing the same issue, One thought I had would be to do the check on the .meta file? as that should only update when the user has logged in and attached their profile?

**Comment by StevenNoel** [comment]
yeah, or ntuser.dat or something like that.

**Comment by RyanSwanepoel** [comment]
For ntuser.dat you would still need to mount the .vhdx to check the time stamp through no? Wouldn't mounting it still change the modified stamp?

**Comment by StevenNoel** [comment]
Looks like the meta file is only for the cloud cache feature.  If it's outside the VHDX file then yeah, that would be ideal.  For customers that aren't using cloud cache, they might not have a choice other than to mount.

**Comment by StevenNoel** [comment]
MrFly had a good suggestion above, that would work.

Or you could do a get-childitem on the meta file, to grab the lastAccessTime and LastWriteTime values. Then do a condition in the code below if the meta file exists, use that files LastAccessTime and LasteWriteTime instead of the VHDX file.

        #If it's older than x days delete disk
        If ( $DeleteOlderThanDays ) {
            #Last Access time isn't always reliable if diff disks are used so lets be safe and use the most recent of access and write
            $mostRecent = $Disk.LastAccessTime, $Disk.LastWriteTime | Measure-Object -Maximum | Select-Object -ExpandProperty Maximum
            if ($mostRecent -lt (Get-Date).AddDays(-$DeleteOlderThanDays) ) {
                try {
                    Remove-Item $Disk.FullName -ErrorAction Stop -Force
                    Write-VhdOutput -DiskState "Deleted" -FinalSize 0 -EndTime (Get-Date)
                }
                catch {
                    Write-VhdOutput -DiskState 'Disk Deletion Failed' -EndTime (Get-Date)
                }
                return
            }
        }

**Comment by RyanSwanepoel** [comment]
Yeah I was thinking about using it but I don't know where to add the Get-Childitem to find the meta file. Will do some more digging and see where I can put it.

**Comment by JimMoyle** [comment]
If you look into the development branch, I've written code there to account
fo this which you can use.

Jim

**Comment by RyanSwanepoel** [comment]
Hi Jim thanks for the response!

Am I correct in thinking I would just need to add the changes in commit (5dc8aa71a55fed418ab1a924ff01f96c1aee9d5b) to both Dismount-FSLDisk.ps1 and Optimize-OneDisk.ps1? which will then set the LastWriteTimeUTC once the disk is unmounted?

Does it need a -SetLastWriteTime switch when calling the script after making the changes?

**Comment by JimMoyle** [comment]
Yep that should do it.

**Comment by JimMoyle** [comment]
I decided against the addition of a switch for the feature for reasons of
simplicity, but feel free to put it in your script :)

Jim

**Comment by RyanSwanepoel** [comment]
Thanks so much Jim, finally got some time to get it working with the changes in the Dev branch.

One quick thought, with the "DeleteOlderThan" option it is only removing the disk itself and leaving behind the meta file and folder structure. Is this by design? 

Can you see any issue with remove the entire folder? so for example changing from Remove-Item $Disk.FullName to Remove-Item $Disk.Path as this would clean up the entire profile (Disks and meta).

**Comment by lordjeb** [comment]
Many configurations will have Profiles and ODFC container vhds in the same directory, so I would not think it is always safe to just delete the directory. The meta file is a more recent addition in non-ccd configurations, and should be cleaned up when the vhd is cleaned up.

**Comment by JimMoyle** [comment]
Nope, can't see an issue with removing the folder, it's a good idea I
should implement

Jim

**Comment by UP932296** [comment]
Can/Has this been implemented?

**Comment by amirjs** [comment]
As MrFly said above, both lastaccesstime and lastwritetime have to be reset after the disk is dismounted. otherwise, the olderthan logic which picks the "maximum" between the two dates will always pick up the most recent i.e. LastAccessTime over LastWriteTime. 

I also had to feed each call to DisMount-FslDisk function with these switches. i.e. 
`DisMount-FslDisk -SetLastWriteTime $lastWriteTime -SetLastAccessTime $lastAccessTime`

Great work and effort Jim - thanks much

**Comment by ashwin2101** [comment]
Hi can someone please share the script? :)

**Comment by Jordyvdp** [comment]
> Thanks so much Jim, finally got some time to get it working with the changes in the Dev branch.
> 
> One quick thought, with the "DeleteOlderThan" option it is only removing the disk itself and leaving behind the meta file and folder structure. Is this by design?
> 
> Can you see any issue with remove the entire folder? so for example changing from Remove-Item $Disk.FullName to Remove-Item $Disk.DirectoryName as this would clean up the entire profile (Disks and meta).

Would it be possible to share the changes you made? I made the changes from the dev branch but the Date modified still changes.

Thank you in advance!
